File: RetailCalcPro/RetailCalcPro/ViewModels/CalculatorViewModel.swift
import Foundation
import SwiftData

@MainActor
@Observable
final class CalculatorViewModel {
    var state = CalculationState()
    var showPaywallSheet = false
    var showSavedFeedback = false

    private var modelContext: ModelContext?
    private let usageLimiter = UsageLimiter.shared

    var isLimitReached: Bool { usageLimiter.isLimitReached }
    var showPaywall: Bool {
        get { usageLimiter.showPaywall }
        set { usageLimiter.showPaywall = newValue }
    }
    var remainingCount: Int { usageLimiter.remainingCount }
    var todayUsageCount: Int { usageLimiter.todayUsageCount }

    func setModelContext(_ context: ModelContext) {
        self.modelContext = context
    }

    // MARK: - Keypad Actions

    func tapDigit(_ digit: String) {
        if state.inputPrice == 0 && digit != "0" {
            guard usageLimiter.canUse() else {
                showPaywallSheet = true
                return
            }
        }
        state.appendDigit(digit)
    }

    func tapDoubleZero() {
        if state.inputPrice == 0 {
            guard usageLimiter.canUse() else {
                showPaywallSheet = true
                return
            }
        }
        state.appendDoubleZero()
    }

    func tapBackspace() {
        state.deleteLastDigit()
    }

    func tapClear() {
        state.clear()
        usageLimiter.recordUsage()
    }

    // MARK: - Tax Actions

    func setTaxRate(_ rate: TaxRate) {
        state.taxRate = rate
    }

    func setTaxMethod(_ method: TaxMethod) {
        state.taxMethod = method
    }

    // MARK: - Discount Actions

    func applyPercentageDiscount(_ percentage: Int) {
        state.setDiscount(.percentage(percentage))
    }

    func clearDiscounts() {
        state.clearDiscounts()
    }

    // MARK: - History（手動保存のみ）

    func saveToHistory() {
        guard let modelContext, state.inputPrice > 0 else {
            print("[Calculator] 保存スキップ: modelContext=\(modelContext != nil), inputPrice=\(state.inputPrice)")
            return
        }
        let result = state.result
        print("[Calculator] 保存開始:")
        print("  入力金額: ¥\(result.inputPrice)")
        print("  割引額: ¥\(result.totalDiscount)")
        print("  割引後: ¥\(result.priceAfterDiscount)")
        print("  税額: ¥\(result.taxAmount) (\(result.taxRate.label), \(result.taxMethod == .exclusive ? "税抜" : "税込"))")
        print("  最終金額: ¥\(result.finalPrice)")

        let history = CalculationHistory(
            inputPrice: result.inputPrice,
            finalPrice: result.finalPrice,
            priceAfterDiscount: result.priceAfterDiscount,
            taxAmount: result.taxAmount,
            totalDiscount: result.totalDiscount,
            taxRate: result.taxRate,
            taxMethod: result.taxMethod,
            discounts: result.discounts
        )
        modelContext.insert(history)
        try? modelContext.save()
        showSavedFeedback = true
        print("[Calculator] 履歴保存完了")
    }

    func restoreFromHistory(_ history: CalculationHistory) {
        state.restore(from: history)
    }
}

File: android/app/src/main/java/com/retailcalcpro/app/viewmodel/CalculatorViewModel.kt
package com.retailcalcpro.app.viewmodel

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.retailcalcpro.app.data.AppDatabase
import com.retailcalcpro.app.data.CalculationHistory
import com.retailcalcpro.app.model.*
import com.retailcalcpro.app.util.TaxCalculator
import com.retailcalcpro.app.util.UsageLimiter
import com.retailcalcpro.app.util.PurchaseManager
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class CalculatorViewModel(application: Application) : AndroidViewModel(application) {

    private val db = AppDatabase.getInstance(application)
    private val historyDao = db.historyDao()
    private val usageLimiter = UsageLimiter.getInstance(application)

    var inputText by mutableStateOf("0")
        private set

    var taxRate by mutableStateOf(TaxRate.STANDARD)
        private set

    var taxMethod by mutableStateOf(TaxMethod.EXCLUSIVE)
        private set

    var appliedDiscounts by mutableStateOf<List<DiscountType>>(emptyList())
        private set

    var showPaywall by mutableStateOf(false)
        private set

    var showSavedFeedback by mutableStateOf(false)
        private set

    val inputPrice: Int
        get() = inputText.toIntOrNull() ?: 0

    val result: CalculationResult
        get() = TaxCalculator.calculate(
            inputPrice = inputPrice,
            taxRate = taxRate,
            taxMethod = taxMethod,
            discounts = appliedDiscounts
        )

    val remainingCount: Int
        get() = usageLimiter.remainingCount

    val isProUnlocked: Boolean
        get() = PurchaseManager.getInstance().isProUnlocked

    val histories = historyDao.getAllHistory()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    fun appendDigit(digit: String) {
        if (!usageLimiter.canUse()) {
            showPaywall = true
            return
        }
        if (inputText.length >= 10) return
        inputText = if (inputText == "0") {
            if (digit == "0") "0" else digit
        } else {
            inputText + digit
        }
    }

    fun appendDoubleZero() {
        if (!usageLimiter.canUse()) {
            showPaywall = true
            return
        }
        if (inputText == "0") return
        if (inputText.length >= 9) return
        inputText += "00"
    }

    fun deleteLastDigit() {
        inputText = if (inputText.length <= 1) "0" else inputText.dropLast(1)
    }

    fun clear() {
        if (inputText != "0") {
            usageLimiter.recordUsage()
        }
        inputText = "0"
    }

    fun updateTaxRate(rate: TaxRate) {
        taxRate = rate
    }

    fun updateTaxMethod(method: TaxMethod) {
        taxMethod = method
    }

    fun applyPercentageDiscount(percentage: Int) {
        appliedDiscounts = listOf(DiscountType.Percentage(percentage))
    }

    fun clearDiscounts() {
        appliedDiscounts = emptyList()
    }

    fun saveToHistory() {
        val r = result
        if (r.inputPrice == 0) return

        viewModelScope.launch {
            val discountPct = when (val d = appliedDiscounts.firstOrNull()) {
                is DiscountType.Percentage -> d.value
                is DiscountType.Amount -> 0
                null -> 0
            }
            historyDao.insert(
                CalculationHistory(
                    inputPrice = r.inputPrice,
                    finalPrice = r.finalPrice,
                    priceAfterDiscount = r.priceAfterDiscount,
                    taxAmount = r.taxAmount,
                    totalDiscount = r.totalDiscount,
                    taxRateValue = when (r.taxRate) {
                        TaxRate.STANDARD -> 10
                        TaxRate.REDUCED -> 8
                        TaxRate.NONE -> 0
                    },
                    taxMethodRaw = when (r.taxMethod) {
                        TaxMethod.EXCLUSIVE -> "exclusive"
                        TaxMethod.INCLUSIVE -> "inclusive"
                    },
                    discountPercent = discountPct
                )
            )
            showSavedFeedback = true
        }
    }

    fun deleteHistory(history: CalculationHistory) {
        viewModelScope.launch {
            historyDao.delete(history)
        }
    }

    fun deleteAllHistory() {
        viewModelScope.launch {
            historyDao.deleteAll()
        }
    }

    fun restoreFromHistory(history: CalculationHistory) {
        inputText = history.inputPrice.toString()
        taxRate = when (history.taxRateValue) {
            10 -> TaxRate.STANDARD
            8 -> TaxRate.REDUCED
            else -> TaxRate.NONE
        }
        taxMethod = if (history.taxMethodRaw == "exclusive") TaxMethod.EXCLUSIVE else TaxMethod.INCLUSIVE
        if (history.discountPercent > 0) {
            appliedDiscounts = listOf(DiscountType.Percentage(history.discountPercent))
        } else {
            appliedDiscounts = emptyList()
        }
    }

    fun dismissPaywall() {
        showPaywall = false
    }

    fun onPurchased() {
        showPaywall = false
    }

    fun dismissSavedFeedback() {
        showSavedFeedback = false
    }
}
